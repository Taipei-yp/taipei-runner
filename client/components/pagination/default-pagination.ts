import { useCallback, useMemo, useState } from "react";

const canGoPrev = (currentPage: number): boolean => currentPage > 1;

const canGoNext = (currentPage: number, allPagesCount: number): boolean =>
  currentPage < allPagesCount;

const getVisibleRange = (
  allPagesCount: number,
  currentPage: number,
  visiblePagesCount: number,
) => {
  const minFirst = 1;
  const maxLast = allPagesCount;

  let first = currentPage - Math.floor(visiblePagesCount / 2);
  let last = first + visiblePagesCount - 1;

  if (first < minFirst) {
    last += minFirst - first;
    first = minFirst;
  } else if (last > maxLast) {
    first -= last - maxLast;
    last = maxLast;
  }

  return [first, last];
};

export const useDefaultPagination = (
  allPagesCount: number,
  suggestedVisiblePagesCount = allPagesCount,
  initialPage = 1,
) => {
  const [currentPage, setCurrentPage] = useState(initialPage);

  const visiblePagesCount = useMemo(() => {
    return Math.min(allPagesCount, suggestedVisiblePagesCount);
  }, [allPagesCount, suggestedVisiblePagesCount]);

  const [firstVisible, lastVisible] = getVisibleRange(
    allPagesCount,
    currentPage,
    visiblePagesCount,
  );

  const canGoToPrevPage = canGoPrev(currentPage);
  const canGoToNextPage = canGoNext(currentPage, allPagesCount);

  const goToPrevPage = useCallback(() => {
    setCurrentPage(currPage => {
      if (canGoPrev(currPage)) {
        return currPage - 1;
      }
      return currPage;
    });
  }, []);

  const goToNextPage = useCallback(() => {
    setCurrentPage(currPage => {
      if (canGoNext(currPage, allPagesCount)) {
        return currPage + 1;
      }
      return currPage;
    });
  }, [allPagesCount]);

  return {
    currentPage,
    firstVisiblePage: firstVisible,
    lastVisiblePage: lastVisible,
    canGoToPrevPage,
    canGoToNextPage,
    goToPage: setCurrentPage,
    goToPrevPage,
    goToNextPage,
  };
};
