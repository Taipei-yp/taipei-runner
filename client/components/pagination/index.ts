export { Pagination } from "./pagination";
export { useDefaultPagination } from "./default-pagination";
