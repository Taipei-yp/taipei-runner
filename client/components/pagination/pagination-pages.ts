import { useCallback, useMemo, useState } from "react";

export function useDefaultPagination(
  allPagesCount: number,
  suggestedVisiblePagesCount = allPagesCount,
  initialPage = 1,
) {
  const [currentPage, setCurrentPage] = useState(initialPage);

  const visiblePagesCount = useMemo(() => {
    return Math.min(allPagesCount, suggestedVisiblePagesCount);
  }, [allPagesCount, suggestedVisiblePagesCount]);

  const [firstVisible, lastVisible] = getVisibleRange(
    allPagesCount,
    currentPage,
    visiblePagesCount,
  );

  const canGoToPrevPage = canGoPrev(currentPage);
  const canGoToNextPage = canGoNext(currentPage, allPagesCount);

  const goToPrevPage = useCallback(() => {
    setCurrentPage(currPage => {
      if (canGoPrev(currPage)) {
        return currPage - 1;
      }
      return currPage;
    });
  }, []);

  const goToNextPage = useCallback(() => {
    setCurrentPage(currPage => {
      if (canGoNext(currPage, allPagesCount)) {
        return currPage + 1;
      }
      return currPage;
    });
  }, [allPagesCount]);

  return {
    currentPage,
    firstVisiblePage: firstVisible,
    lastVisiblePage: lastVisible,
    canGoToPrevPage,
    canGoToNextPage,
    goToPage: setCurrentPage,
    goToPrevPage,
    goToNextPage,
  };
}

function getVisibleRange(
  allPagesCount: number,
  currentPage: number,
  visiblePagesCount: number,
) {
  const minFirst = 1;
  const maxLast = allPagesCount;

  let first = currentPage - Math.floor(visiblePagesCount / 2);
  let last = first + visiblePagesCount - 1;

  if (first < minFirst) {
    last += minFirst - first;
    first = minFirst;
  } else if (last > maxLast) {
    first -= last - maxLast;
    last = maxLast;
  }

  return [first, last];
}

function canGoPrev(currentPage: number): boolean {
  return currentPage > 1;
}

function canGoNext(currentPage: number, allPagesCount: number): boolean {
  return currentPage < allPagesCount;
}
