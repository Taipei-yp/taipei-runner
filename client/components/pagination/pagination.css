.pagination {
  margin: 0;
  padding: 0;
  list-style: none;
  color: var(--tpr-color-secondary);
  font-size: var(--tpr-fontsize-l);
  user-select: none;

  &__item {
    display: inline-block;
    margin: 0 0.4em;
    cursor: pointer;

    &_active {
      color: var(--tpr-color-primary);
    }

    &_disabled {
      color: var(--tpr-color-dark-deep);
      cursor: initial;
    }
  }
}
