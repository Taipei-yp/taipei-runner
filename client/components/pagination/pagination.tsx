import React, { FC, memo, useCallback, MouseEvent } from "react";
import block from "bem-cn";
import "./pagination.css";
import { arrayFromTo } from "../../utils/array-utils";

const b = block("pagination");

type Props = {
  className?: string;
  firstPage: number;
  lastPage: number;
  currentPage: number;
  canGoToNextPage: boolean;
  canGoToPrevPage: boolean;
  onPageClick: (pageNumber: number) => void;
  onGoToPrevPageClick?: () => void;
  onGoToNextPageClick?: () => void;
};

const Pagination: FC<Props> = ({
  className,
  firstPage,
  lastPage,
  currentPage,
  canGoToNextPage,
  canGoToPrevPage,
  onPageClick,
  onGoToPrevPageClick,
  onGoToNextPageClick,
}) => {
  const handleClick = useCallback(
    (event: MouseEvent<HTMLLIElement>) => {
      const { pageNumber } = event.currentTarget.dataset;
      onPageClick(Number.parseInt(pageNumber!, 10));
    },
    [onPageClick],
  );

  return (
    <ul className={b.mix(className)}>
      <li
        className={b("item", { disabled: !canGoToPrevPage })}
        onClick={onGoToPrevPageClick}
      >
        {"<"}
      </li>
      {arrayFromTo(firstPage, lastPage).map(page => (
        <li
          key={page}
          className={b("item", { active: page === currentPage })}
          data-page-number={page}
          onClick={handleClick}
        >
          {page}
        </li>
      ))}
      <li
        className={b("item", { disabled: !canGoToNextPage })}
        onClick={onGoToNextPageClick}
      >
        {">"}
      </li>
    </ul>
  );
};

const WrappedPagination = memo(Pagination);
export { WrappedPagination as Pagination };
