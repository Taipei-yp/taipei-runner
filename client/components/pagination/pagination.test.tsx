import React from "react";
import TestRenderer from "react-test-renderer";
import { Pagination } from "client/components/pagination/pagination";

describe("Component Pagination", () => {
  it("renders all elements", () => {
    const component = TestRenderer.create(
      <Pagination
        firstPage={1}
        lastPage={5}
        currentPage={2}
        canGoToNextPage={false}
        canGoToPrevPage={false}
        onPageClick={() => {}}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("shows next/prev buttons as disabled when can't go next/prev", () => {
    const component = TestRenderer.create(
      <Pagination
        firstPage={1}
        lastPage={3}
        currentPage={1}
        canGoToNextPage={false}
        canGoToPrevPage={false}
        onPageClick={() => {}}
      />,
    );
    const listItems = component.root.findAllByType("li", { deep: true });
    const goPrevButton = listItems[0];
    const goNextButton = listItems[listItems.length - 1];
    expect(goPrevButton.props.className.toString()).toContain("disabled");
    expect(goNextButton.props.className.toString()).toContain("disabled");
  });

  it("shows next/prev buttons as enabled when can go next/prev", () => {
    const component = TestRenderer.create(
      <Pagination
        firstPage={1}
        lastPage={3}
        currentPage={1}
        canGoToNextPage
        canGoToPrevPage
        onPageClick={() => {}}
      />,
    );
    const listItems = component.root.findAllByType("li", { deep: true });
    const goPrevButton = listItems[0];
    const goNextButton = listItems[listItems.length - 1];
    expect(goPrevButton.props.className.toString()).not.toContain("disabled");
    expect(goNextButton.props.className.toString()).not.toContain("disabled");
  });

  it("marks current page as active", () => {
    const component = TestRenderer.create(
      <Pagination
        firstPage={1}
        lastPage={3}
        currentPage={1}
        canGoToNextPage
        canGoToPrevPage
        onPageClick={() => {}}
      />,
    );

    let [
      ,
      firstPage,
      secondPage,
      thirdPage,
    ] = component.root.findAllByType("li", { deep: true });
    expect(firstPage.props.className.toString()).toContain("active");
    expect(secondPage.props.className.toString()).not.toContain("active");
    expect(thirdPage.props.className.toString()).not.toContain("active");

    TestRenderer.act(() => {
      component.update(
        <Pagination
          firstPage={1}
          lastPage={3}
          currentPage={2}
          canGoToNextPage
          canGoToPrevPage
          onPageClick={() => {}}
        />,
      );
    });

    [, firstPage, secondPage, thirdPage] = component.root.findAllByType("li", {
      deep: true,
    });
    expect(firstPage.props.className.toString()).not.toContain("active");
    expect(secondPage.props.className.toString()).toContain("active");
    expect(thirdPage.props.className.toString()).not.toContain("active");
  });
});
