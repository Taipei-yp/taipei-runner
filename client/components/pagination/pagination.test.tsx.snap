// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Pagination renders all elements 1`] = `
<ul
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <li
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    &lt;
  </li>
  <li
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    data-page-number={1}
    onClick={[Function]}
  >
    1
  </li>
  <li
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    data-page-number={2}
    onClick={[Function]}
  >
    2
  </li>
  <li
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    data-page-number={3}
    onClick={[Function]}
  >
    3
  </li>
  <li
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    data-page-number={4}
    onClick={[Function]}
  >
    4
  </li>
  <li
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
    data-page-number={5}
    onClick={[Function]}
  >
    5
  </li>
  <li
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    &gt;
  </li>
</ul>
`;
